rust/connlib/tunnel/src/client.rs

```diff
@@ -160,48 +165,45 @@ where
         Ok(())
     }
 
-    /// Sets the interface configuration and starts background tasks.
-    #[tracing::instrument(level = "trace", skip(self))]
-    pub fn set_interface(
-        &mut self,
-        config: &InterfaceConfig,
-        dns_mapping: BiMap<IpAddr, DnsServer>,
-    ) -> connlib_shared::Result<()> {
-        let device = Device::new(
-            config,
+    pub(crate) fn update_interface(&mut self) -> connlib_shared::Result<()> {
+        let dns_mapping = self.role_state.dns_mapping();
+        let mut device = Device::new(
+            self.role_state.interface_config.as_ref().expect("Developer error: we should always call update_interface after the interface config is set"),
             // We can just sort in here because sentinel ips are created in order
             dns_mapping.left_values().copied().sorted().collect(),
             self.callbacks(),
         )?;
 
+        device.set_routes(self.role_state.routes().collect(), &self.callbacks)?;
+
         self.device = Some(device);
         self.no_device_waker.wake();
 
-        let mut errs = Vec::new();
-        for sentinel in dns_mapping.left_values() {
-            if let Err(e) = self.add_route((*sentinel).into()) {
-                tracing::warn!(err = ?e, %sentinel , "couldn't add route for sentinel");
-                errs.push(e);
-            }
-        }
-
-        if errs.len() == dns_mapping.left_values().len() && dns_mapping.left_values().len() > 0 {
-            return Err(errs.pop().unwrap());
-        }
-
-        self.role_state.set_dns_mapping(dns_mapping);
-
-        let res_v4 = self.add_route(IPV4_RESOURCES.parse().unwrap());
-        let res_v6 = self.add_route(IPV6_RESOURCES.parse().unwrap());
-        res_v4.or(res_v6)?;
-
         self.callbacks.on_tunnel_ready()?;
 
         tracing::debug!("background_loop_started");
 
         Ok(())
     }
 
+    /// Sets the interface configuration and starts background tasks.
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn set_interface(&mut self, config: &InterfaceConfig) -> connlib_shared::Result<()> {
+        self.role_state.interface_config = Some(config.clone());
+        let effective_dns_servers = effective_dns_servers(
+            config.upstream_dns.clone(),
+            self.callbacks()
+                .get_system_default_resolvers()
+                .ok()
```

Comment:
Same as for the other callback: If we can't do anything meaningful with the error, I'd suggest not returning an error in the first place and force the caller to log and give us an empty list instead.